Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -405,23 +415,53 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
         RefPtr strongThis = weakThis.get();
         if (!strongThis)
             return;
-        if (error)
-            return; // FIXME: Pipe this error to JS.
+        if (error) {
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                strongThis->receiveDatagram(std::span<const uint8_t>(), false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
+            return;
+        }
+
+        ASSERT(content || withFin);
 
         // FIXME: Not only is this an unnecessary string copy, but it's also something that should probably be in WTF or FragmentedSharedBuffer.
         auto vectorFromData = [](dispatch_data_t content) {
-            ASSERT(content);
             Vector<uint8_t> request;
-            dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
-                request.append(buffer);
-                return true;
-            });
+            if (content) {
+                dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
+                    request.append(buffer);
+                    return true;
+                });
+            }
             return request;
         };
 
-        strongThis->receiveDatagram(vectorFromData(content).span());
-        strongThis->receiveDatagramLoop();
+        bool completed = !content && withFin;
+        strongThis->receiveDatagram(vectorFromData(content).span(), completed, std::nullopt);
+        if (!completed)
+            strongThis->receiveDatagramLoop();
     }).get());
 #endif // HAVE(WEB_TRANSPORT)
 }
+
+void NetworkTransportSession::terminate(uint32_t code, CString&& message)
+{
+#if HAVE(WEB_TRANSPORT)
+    RetainPtr metadata = adoptNS((nw_protocol_metadata_t)nw_retain(nw_connection_group_copy_protocol_metadata(m_connectionGroup.get(), nw_protocol_copy_webtransport_definition())));
+    if (metadata) {
```

Comment:
if (RetainPtr metadata = adoptNS(...))
Then metadata has the right scope and you don't need to explicitly set it to null.  Also, the nw_retain call looks like it'll cause a memory leak.